### DISCUSSION TITLE
"Failed to establish connection to NetBox API" - again

### DISCUSSION THREAD

**Discussion by mauri5hut** [netbox-community/ansible_modules#822] [Q&A] [answered]
I'm running into this well known error with the following environment:

----------------------------------
Ansible Controller (Debian 11):

- ansible: 2.10.8
- python: 3.9.2
- pynetbox: 5.3.1-1 (6.x also does not work)
- netbox.netbox Collection: 3.7.1

Netbox Server (FreeBSD Jail 13.1):
   
- netbox: 3.2.8
----------------------------------

Playbook:

```
- name: Running netbox tasks
  connection: local
  hosts: localhost
  gather_facts: no

  tasks:

    - name: Create new child prefix
      netbox.netbox.netbox_prefix:
        netbox_url: "{{ url }}/api/ipam/prefixes/"
        netbox_token: "{{ authToken }}"
        validate_certs: /usr/share/ca-certificates/Root-CA.pem
        data:
          parent: 10.21.0.0/16
          prefix_length: 24
        state: present
        first_available: yes
```

It doesn't matter which module I will use, the error remains the same.

The debuged error (with -vvv):

```
The full traceback is:
  File "/tmp/ansible_netbox.netbox.netbox_prefix_payload_uyfqgby5/ansible_netbox.netbox.netbox_prefix_payload.zip/ansible_collections/netbox/netbox/plugins/module_utils/netbox_utils.py", line 680, in _connect_netbox_api
    self.version = nb.version
  File "/usr/lib/python3/dist-packages/pynetbox/core/api.py", line 127, in version
    version = Request(
  File "/usr/lib/python3/dist-packages/pynetbox/core/query.py", line 193, in get_version
    raise RequestError(req)
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "cert": null,
            "data": {
                "custom_fields": null,
                "description": null,
                "family": null,
                "is_pool": null,
                "parent": "10.21.0.0/16",
                "prefix": null,
                "prefix_length": 24,
                "prefix_role": null,
                "site": null,
                "status": null,
                "tags": null,
                "tenant": null,
                "vlan": null,
                "vrf": null
            },
            "first_available": true,
            "netbox_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "netbox_url": "https://netbox.mobilelab.local/api/ipam/prefixes/",
            "query_params": null,
            "state": "present",
            "validate_certs": "/usr/share/ca-certificates/Root-CA.pem"
        }
    },
    "msg": "Failed to establish connection to NetBox API"
```

The token is the generated token, which also works by REST (with Insomnia).
The used module is: /root/.ansible/collections/ansible_collections/netbox/netbox/plugins/modules/netbox_prefix.py

What seems strange to me, is this part in the error message:

```
  File "/usr/lib/python3/dist-packages/pynetbox/core/query.py", line 193, in get_version
    raise RequestError(req)
fatal: [localhost]: FAILED! => {
```

This looks like, if the module is trying to get the netbox version from the localhost - which is the ansible controller.
Is a remote execution possible? Have I misunderstood something here?

### Side topic:

**About the "validate_certs" option in the playbook:** 

Very strange. The Netbox-URL is reachable by HTTPS, no certification errors (wether if checked with a Webbrowser, or the openssl command). It is also possible to do all things by REST with Insomnia. The certificate is a valid PKI certificate, the chain is also valid and the CA certificate is installed systemwide. So I have no idea, why the module needs this option explicit with the path to the certificate.
If I don't use the option, the error is following:

```
The full traceback is:
  File "/tmp/ansible_netbox.netbox.netbox_prefix_payload_98v993wf/ansible_netbox.netbox.netbox_prefix_payload.zip/ansible_collections/netbox/netbox/plugins/module_utils/netbox_utils.py", line 680, in _connect_netbox_api
    self.version = nb.version
  File "/usr/lib/python3/dist-packages/pynetbox/core/api.py", line 127, in version
    version = Request(
  File "/usr/lib/python3/dist-packages/pynetbox/core/query.py", line 189, in get_version
    req = self.http_session.get(self.normalize_url(self.base), headers=headers,)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 555, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 514, in send
    raise SSLError(e, request=request)
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "cert": null,
            "data": {
                "custom_fields": null,
                "description": null,
                "family": null,
                "is_pool": null,
                "parent": "10.21.0.0/16",
                "prefix": null,
                "prefix_length": 24,
                "prefix_role": null,
                "site": null,
                "status": null,
                "tags": null,
                "tenant": null,
                "vlan": null,
                "vrf": null
            },
            "first_available": true,
            "netbox_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "netbox_url": "https://netbox.mobilelab.local/api/ipam/prefixes/",
            "query_params": null,
            "state": "present",
            "validate_certs": true
        }
    },
    "msg": "Failed to establish connection to NetBox API"
```

**Comment by mauri5hut** [comment]
**Update:** No changes with Ansible v2.13.3 and also the same behaviour with another Netbox Server hosted on Debian 11.

**Comment by mauri5hut** [comment]
**Update2:** Same error with Netbox 3.3.2 on Debian 11.

**Reply by rodvand** [reply]
Are you running a multiple redis setup? Can you go to your netbox and get the output from `https://<YOUR_NETBOX>/api/status/`?

**Reply by mauri5hut** [reply]
Original Setup (FreeBSD):

`
HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "django-version": "4.0.7",
    "installed-apps": {
        "debug_toolbar": "3.5.0",
        "django_filters": "22.1",
        "django_prometheus": "2.2.0",
        "django_rq": "2.5.1",
        "django_tables2": "2.4.1",
        "drf_yasg": "1.20.0",
        "graphene_django": "2.15.0",
        "graphiql_debug_toolbar": "0.2.0",
        "mptt": "0.13.4",
        "rest_framework": "3.13.1",
        "social_django": "5.0.0",
        "taggit": "2.1.0",
        "timezone_field": "5.0"
    },
    "netbox-version": "3.2.8",
    "plugins": {},
    "python-version": "3.9.13",
    "rq-workers-running": 2
}
`

I'm not running a multiple redis setup.

`
[root@jlntbx /]# ps faux | grep redis
redis    24773  0.0  0.3  41312  11908  -  SsJ  Wed07   1:48.04 redis-server: /usr/local/bin/redis-server 127.0.0.1:6379 (redis-server)
`

**Comment by mauri5hut** [comment, marked as answer]
I ended up writing the Playbooks/Tasks with the built in uri module. With this everything works as expected.

**Comment by MrTrinax** [comment]
Experiencing the same issue with identical error message on Netbox 3.21.0 and Ansible core 2.16.3 on Ubuntu 24.04.
I ended up using URI instead, but it's a shame not being able to use the Netbox modules
